File: lib/classifier.js
// ═══════════════════════════════════════════════════════════
// 분류기 — 저장 형식과 추론 (TF.js 없이 도는 순수 JS)
// ═══════════════════════════════════════════════════════════
// 학습은 TF.js(lib/trainer.js)가 하지만, 저장·추론은 TF.js 를 쓰지 않는다.
// 가중치를 평범한 float32 덩어리로 내보내야 파이썬에서도 numpy 로 바로 읽는다.
//
// 저장 형식 (두 파일)
//   classifier.json  구조 설명 — 층마다 kernel/bias 의 shape 과 offset(float32 개수)
//   classifier.bin   float32 little-endian 을 순서대로 이어 붙인 것
//
// 층 구성: 입력 D → dense(relu) → dense(softmax). dropout 은 학습에만 쓰이므로
// 저장하지 않는다.

export const CLF_FORMAT = 'teachlab-classifier';
export const CLF_VERSION = 1;

// ── TF.js 모델 → { json, bin } ──
export async function serialize(model, classes, inputDim) {
  const layers = [];
  const chunks = [];
  let offset = 0;

  const dense = model.layers.filter(l => l.getClassName() === 'Dense');
  for (let i = 0; i < dense.length; i++) {
    const [k, b] = dense[i].getWeights();
    const kd = await k.data(), bd = await b.data();
    const kShape = k.shape.slice(), bShape = b.shape.slice();
    layers.push({
      type: 'dense',
      activation: i === dense.length - 1 ? 'softmax' : 'relu',
      kernel: { shape: kShape, offset, count: kd.length },
      bias: { shape: bShape, offset: offset + kd.length, count: bd.length },
    });
    chunks.push(Float32Array.from(kd), Float32Array.from(bd));
    offset += kd.length + bd.length;
  }

  const bin = new Float32Array(offset);
  let at = 0;
  chunks.forEach(c => { bin.set(c, at); at += c.length; });

  return {
    json: {
      format: CLF_FORMAT,
      version: CLF_VERSION,
      inputDim,
      classes: classes.slice(),
      dtype: 'float32-le',
      weightsFile: 'classifier.bin',
      totalCount: offset,
      layers,
    },
    bin: bin.buffer,
  };
}

// ── { json, bin } → 추론 가능한 분류기 ──
export function deserialize(json, binBuffer) {
  if (!json || json.format !== CLF_FORMAT) throw new Error('not a teachlab classifier');
  const all = new Float32Array(binBuffer);
  if (all.length < json.totalCount) throw new Error('weights file is too short');
  const layers = json.layers.map(l => ({
    activation: l.activation,
    inDim: l.kernel.shape[0],
    outDim: l.kernel.shape[1],
    w: all.subarray(l.kernel.offset, l.kernel.offset + l.kernel.count),
    b: all.subarray(l.bias.offset, l.bias.offset + l.bias.count),
  }));
  return { inputDim: json.inputDim, classes: json.classes.slice(), layers };
}

// ── 한 장 추론 → 확률 Float32Array ──
export function predict(clf, vec) {
  let x = vec;
  for (const l of clf.layers) {
    const out = new Float32Array(l.outDim);
    for (let j = 0; j < l.outDim; j++) {
      let s = l.b[j];
      for (let i = 0; i < l.inDim; i++) s += x[i] * l.w[i * l.outDim + j];
      out[j] = s;
    }
    if (l.activation === 'relu') {
      for (let j = 0; j < out.length; j++) if (out[j] < 0) out[j] = 0;
    } else if (l.activation === 'softmax') {
      let mx = -Infinity;
      for (let j = 0; j < out.length; j++) if (out[j] > mx) mx = out[j];
      let sum = 0;
      for (let j = 0; j < out.length; j++) { out[j] = Math.exp(out[j] - mx); sum += out[j]; }
      for (let j = 0; j < out.length; j++) out[j] /= sum || 1;
    }
    x = out;
  }
  return x;
}

// ── 큰 이진 덩어리 ↔ base64 (프로젝트 파일에 담기 위해) ──
// 스택이 터지지 않게 조각내어 변환한다
export function bytesToB64(bytes) {
  let s = '';
  const CH = 0x8000;
  for (let i = 0; i < bytes.length; i += CH) {
    s += String.fromCharCode.apply(null, bytes.subarray(i, i + CH));
  }
  return btoa(s);
}

export function b64ToBytes(b64) {
  const s = atob(b64);
  const out = new Uint8Array(s.length);
  for (let i = 0; i < s.length; i++) out[i] = s.charCodeAt(i);
  return out;
}

File: python/teachlab/classifier.py
# -*- coding: utf-8 -*-
"""분류기 — Teach Lab 이 내보낸 가중치를 numpy 로 그대로 돌린다.

저장 형식 (파일 두 개)
  classifier.json  층마다 kernel/bias 의 shape 과 offset(float32 개수)
  classifier.bin   float32 little-endian 을 순서대로 이어 붙인 것

딥러닝 라이브러리가 필요 없다. 행렬 곱 두 번이면 끝난다.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Sequence

import numpy as np

CLF_FORMAT = "teachlab-classifier"
CLF_VERSION = 1


@dataclass
class Prediction:
    """한 장에 대한 답."""

    label: str
    score: float
    probs: dict[str, float]

    def ranked(self) -> list[tuple[str, float]]:
        """확률이 높은 순서로 [(이름, 확률), ...]"""
        return sorted(self.probs.items(), key=lambda kv: kv[1], reverse=True)

    def __str__(self) -> str:                         # print() 할 때 보기 좋게
        return f"{self.label} ({self.score * 100:.1f}%)"


@dataclass
class _Layer:
    weight: np.ndarray                                # (in, out)
    bias: np.ndarray                                  # (out,)
    activation: str


class Classifier:
    """평범한 다층 퍼셉트론 (Dense relu → Dense softmax)."""

    def __init__(self, spec: dict, weights: bytes):
        if spec.get("format") != CLF_FORMAT:
            raise ValueError("Teach Lab 분류기 파일이 아니에요.")
        if int(spec.get("version", 1)) > CLF_VERSION:
            raise ValueError("더 새로운 버전에서 만든 파일이에요. teachlab 을 올려 주세요.")

        flat = np.frombuffer(weights, dtype="<f4")
        total = int(spec.get("totalCount", flat.size))
        if flat.size < total:
            raise ValueError("가중치 파일이 잘렸어요.")

        self.input_dim = int(spec["inputDim"])
        self.classes: list[str] = list(spec["classes"])
        self.layers: list[_Layer] = []
        for layer in spec["layers"]:
            k, b = layer["kernel"], layer["bias"]
            weight = flat[k["offset"]:k["offset"] + k["count"]].reshape(tuple(k["shape"]))
            bias = flat[b["offset"]:b["offset"] + b["count"]].reshape(tuple(b["shape"]))
            self.layers.append(_Layer(weight, bias, layer.get("activation", "relu")))

    # ── 만들기 ──
    @classmethod
    def from_dir(cls, folder: str | Path) -> "Classifier":
        folder = Path(folder)
        spec = json.loads((folder / "classifier.json").read_text(encoding="utf-8"))
        name = spec.get("weightsFile", "classifier.bin")
        return cls(spec, (folder / name).read_bytes())

    # ── 계산 ──
    def predict_proba(self, vec: Sequence[float] | np.ndarray) -> np.ndarray:
        x = np.asarray(vec, dtype=np.float32).reshape(-1)
        if x.size != self.input_dim:
            raise ValueError(f"입력이 {self.input_dim} 개여야 하는데 {x.size} 개가 왔어요.")
        for layer in self.layers:
            x = x @ layer.weight + layer.bias
            if layer.activation == "relu":
                x = np.maximum(x, 0.0)
            elif layer.activation == "softmax":
                e = np.exp(x - x.max())
                x = e / (e.sum() or 1.0)
        return x.astype(np.float32)

    def predict(self, vec: Sequence[float] | np.ndarray) -> Prediction:
        probs = self.predict_proba(vec)
        best = int(np.argmax(probs))
        return Prediction(
            label=self.classes[best],
            score=float(probs[best]),
            probs={name: float(p) for name, p in zip(self.classes, probs)},
        )
